tests/doc_index/base_classes/test_base_doc_store.py

```diff
@@ -151,81 +151,115 @@ class OtherFlatDoc(FlatDoc):
-                    tens_one=np.random.random((10,)), tens_two=np.random.random((50,))
-                )
-            ]
+    assert store._validate_docs(in_other_da) == in_other_da
+    with pytest.raises(ValueError):
+        store._validate_docs([SimpleDoc(tens=np.random.random((10,)))])
+    with pytest.raises(ValueError):
+        assert not store._validate_docs(
+            DocumentArray[SimpleDoc]([SimpleDoc(tens=np.random.random((10,)))])
         )
-    )
-    assert not store._is_schema_compatible([SimpleDoc(tens=np.random.random((10,)))])
-    assert not store._is_schema_compatible(
-        DocumentArray[SimpleDoc]([SimpleDoc(tens=np.random.random((10,)))])
-    )
 
+    # NESTED
     store = DummyDocIndex[NestedDoc]()
-    assert store._is_schema_compatible(
+    in_list = [NestedDoc(d=SimpleDoc(tens=np.random.random((10,))))]
+    assert isinstance(store._validate_docs(in_list), DocumentArray[BaseDocument])
+    in_da = DocumentArray[NestedDoc](
         [NestedDoc(d=SimpleDoc(tens=np.random.random((10,))))]
     )
-    assert store._is_schema_compatible(
-        DocumentArray[NestedDoc]([NestedDoc(d=SimpleDoc(tens=np.random.random((10,))))])
-    )
-    assert store._is_schema_compatible(
+    assert store._validate_docs(in_da) == in_da
+    in_other_list = [OtherNestedDoc(d=OtherSimpleDoc(tens=np.random.random((10,))))]
+    assert isinstance(store._validate_docs(in_other_list), DocumentArray[BaseDocument])
+    in_other_da = DocumentArray[OtherNestedDoc](
         [OtherNestedDoc(d=OtherSimpleDoc(tens=np.random.random((10,))))]
     )
-    assert store._is_schema_compatible(
-        DocumentArray[OtherNestedDoc](
-            [OtherNestedDoc(d=OtherSimpleDoc(tens=np.random.random((10,))))]
+
+    assert store._validate_docs(in_other_da) == in_other_da
+    with pytest.raises(ValueError):
+        store._validate_docs([SimpleDoc(tens=np.random.random((10,)))])
+    with pytest.raises(ValueError):
+        store._validate_docs(
+            DocumentArray[SimpleDoc]([SimpleDoc(tens=np.random.random((10,)))])
         )
-    )
-    assert not store._is_schema_compatible([SimpleDoc(tens=np.random.random((10,)))])
-    assert not store._is_schema_compatible(
-        DocumentArray[SimpleDoc]([SimpleDoc(tens=np.random.random((10,)))])
-    )
+
+
+def test_docs_validation_unions():
+    class OptionalDoc(BaseDocument):
+        tens: Optional[NdArray[10]] = Field(dim=1000)
+
+    class UnionDoc(BaseDocument):
+        tens: Union[NdArray[10], str] = Field(dim=1000)
+
+    # OPTIONAL
+    store = DummyDocIndex[SimpleDoc]()
+    in_list = [OptionalDoc(tens=np.random.random((10,)))]
+    assert isinstance(store._validate_docs(in_list), DocumentArray[BaseDocument])
+    in_da = DocumentArray[OptionalDoc](in_list)
+    assert store._validate_docs(in_da) == in_da
+
+    with pytest.raises(ValueError):
+        store._validate_docs([OptionalDoc(tens=None)])
+
+    # OTHER UNION
+    store = DummyDocIndex[SimpleDoc]()
```

Comment:
should we add tests for `HnswDocumentIndex` too?

Reply:
I will do that separately, since work on that index is still ongoing